docs/doxygen/man/man3/Mount.3
.TH "Mount" 3 "Thu May 12 2022" "Version 0.2" "libmount-cpp" \" -*- nroff -*-
.ad l
.nh
.SH NAME
Mount
.SH SYNOPSIS
.br
.PP
.SS "Public Types"

.in +1c
.ti -1c
.RI "enum { \fBadfs\fP, \fBaffs\fP, \fBautofs\fP, \fBcifs\fP, \fBcoda\fP, \fBcoherent\fP, \fBcramfs\fP, \fBdebugfs\fP, \fBdevpts\fP, \fBefs\fP, \fBext\fP, \fBext2\fP, \fBext3\fP, \fBext4\fP, \fBhfs\fP, \fBhfsplus\fP, \fBhpfs\fP, \fBiso9660\fP, \fBjfs\fP, \fBminix\fP, \fBmsdos\fP, \fBncpfs\fP, \fBnfs\fP, \fBnfs4\fP, \fBntfs\fP, \fBproc\fP, \fBqnx4\fP, \fBramfs\fP, \fBreiserfs\fP, \fBromfs\fP, \fBsquashfs\fP, \fBsmbfs\fP, \fBsysv\fP, \fBtmpfs\fP, \fBubifs\fP, \fBudf\fP, \fBufs\fP, \fBumsdos\fP, \fBusbfs\fP, \fBvfat\fP, \fBxenix\fP, \fBxfs\fP, \fBxiafs\fP }"
.br
.in -1c
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "int \fBfsMount\fP (std::string fsDir, std::string outDir, int fsType, std::string accessType)"
.br
.RI "Mounts a filesystem\&. "
.ti -1c
.RI "int \fBfsUmount\fP (std::string outDir)"
.br
.RI "Unmounts a mounted directory\&. "
.ti -1c
.RI "int \fBfsUmount\fP ()"
.br
.RI "Unmounts the last mounted directory mounted with the object\&. "
.in -1c
.SS "Static Public Attributes"

.in +1c
.ti -1c
.RI "static constexpr int \fBNOROOT\fP = \-3"
.br
.RI "NOROOT is the error return code for a user trying to use this without root\&. "
.ti -1c
.RI "static constexpr int \fBMOUNTPOINTERR\fP = \-2"
.br
.RI "MOUNTPOINTERR is the error returned code for the mount point not existing\&. "
.ti -1c
.RI "static constexpr int \fBMOUNTFILEERR\fP = \-1"
.br
.RI "MOUNTFILEERR is the error return code for the file system mount not existing\&. "
.ti -1c
.RI "static const std::string \fBREADONLY\fP = 'ro'"
.br
.RI "READONLY is the Read Only attribute\&. "
.ti -1c
.RI "static const std::string \fBREADWRITE\fP = 'rw'"
.br
.RI "READWRITE is the Read and Write attribute\&. "
.in -1c
.SH "Member Enumeration Documentation"
.PP 
.SS "anonymous enum"

.PP
\fBEnumerator\fP
.in +1c
.TP
\fB\fIadfs \fP\fP
adfs 
.TP
\fB\fIaffs \fP\fP
affs 
.TP
\fB\fIautofs \fP\fP
autofs 
.TP
\fB\fIcifs \fP\fP
cifs 
.TP
\fB\fIcoda \fP\fP
coda 
.TP
\fB\fIcoherent \fP\fP
coherent 
.TP
\fB\fIcramfs \fP\fP
cramfs 
.TP
\fB\fIdebugfs \fP\fP
debugfs 
.TP
\fB\fIdevpts \fP\fP
devpts 
.TP
\fB\fIefs \fP\fP
efs 
.TP
\fB\fIext \fP\fP
ext 
.TP
\fB\fIext2 \fP\fP
ext2 
.TP
\fB\fIext3 \fP\fP
ext3 
.TP
\fB\fIext4 \fP\fP
ext4 
.TP
\fB\fIhfs \fP\fP
hfs 
.TP
\fB\fIhfsplus \fP\fP
hfsplus 
.TP
\fB\fIhpfs \fP\fP
hpfs 
.TP
\fB\fIiso9660 \fP\fP
iso9660 
.TP
\fB\fIjfs \fP\fP
jfs 
.TP
\fB\fIminix \fP\fP
minix 
.TP
\fB\fImsdos \fP\fP
msdos 
.TP
\fB\fIncpfs \fP\fP
ncpfs 
.TP
\fB\fInfs \fP\fP
nfs 
.TP
\fB\fInfs4 \fP\fP
nfs4 
.TP
\fB\fIntfs \fP\fP
ntfs 
.TP
\fB\fIproc \fP\fP
proc 
.TP
\fB\fIqnx4 \fP\fP
qnx4 
.TP
\fB\fIramfs \fP\fP
ramfs 
.TP
\fB\fIreiserfs \fP\fP
reiserfs 
.TP
\fB\fIromfs \fP\fP
romfs 
.TP
\fB\fIsquashfs \fP\fP
squashfs 
.TP
\fB\fIsmbfs \fP\fP
smbfs 
.TP
\fB\fIsysv \fP\fP
sysv 
.TP
\fB\fItmpfs \fP\fP
tmpfs 
.TP
\fB\fIubifs \fP\fP
ubifs 
.TP
\fB\fIudf \fP\fP
udf 
.TP
\fB\fIufs \fP\fP
ufs 
.TP
\fB\fIumsdos \fP\fP
umsdos 
.TP
\fB\fIusbfs \fP\fP
usbfs 
.TP
\fB\fIvfat \fP\fP
vfat 
.TP
\fB\fIxenix \fP\fP
xenix 
.TP
\fB\fIxfs \fP\fP
xfs 
.TP
\fB\fIxiafs \fP\fP
xiafs 
.SH "Member Function Documentation"
.PP 
.SS "int Mount::fsMount (std::string fsDir, std::string outDir, int fsType, std::string accessType)"

.PP
Mounts a filesystem\&. 
.PP
\fBParameters\fP
.RS 4
\fIfsDir\fP The filesystem file, ex\&. \fC/dev/sdc\fP 
.br
\fIoutDir\fP The output or mountpoint directory\&. ex\&. \fC/home/user/mountpoint\fP 
.br
\fIfsType\fP The Filesystem type\&. ex\&. 'squashfs' \fC\fBMount::squashfs\fP\fP 
.br
\fIaccessType\fP The access type\&. Read Only or Read Write\&. ex\&. \fC\fBMount::READWRITE\fP\fP \fC\fBMount::READONLY\fP\fP 
.RE
.PP
\fBReturns\fP
.RS 4
int 
.RE
.PP

.SS "int Mount::fsUmount ()"

.PP
Unmounts the last mounted directory mounted with the object\&. 
.PP
\fBReturns\fP
.RS 4
int 
.RE
.PP

.SS "int Mount::fsUmount (std::string outDir)"

.PP
Unmounts a mounted directory\&. 
.PP
\fBParameters\fP
.RS 4
\fIoutDir\fP The directory that is mounted to 
.RE
.PP
\fBReturns\fP
.RS 4
int 
.RE
.PP


.SH "Author"
.PP 
Generated automatically by Doxygen for libmount-cpp from the source code\&.


src/Mount.cpp
/*
 * Mount.cpp
 *
 *  Created on: May 9, 2022
 *      Author: thetimbrick
 */

#include "Mount.h"

#include <filesystem>

const std::string Mount::READONLY = "ro";

const std::string Mount::READWRITE = "rw";

Mount::Mount()

{

}

Mount::~Mount() {

}

int Mount::fsMount(std::string fsDir, std::string outDir, int fsType, std::string accessType) {
	// Checks to see if we are running as root.
	// Root is required to use the `mount` command.
	if (geteuid() != 0) {
		return Mount::NOROOT;
	}
	// Checks to see if the file to mount exists.
	// If not, returns an error.
	if (!std::filesystem::exists(fsDir)) {
		return Mount::MOUNTFILEERR;
	}
	// Checks to see if the mount directory exists.
	// If not, returns an error.
	if (!std::filesystem::exists(outDir)) {
		return Mount::MOUNTPOINTERR;
	}
	// Update mount variables
	m_currentMountFS = Mount::m_getMountString(fsType);
	m_currentMountPoint = outDir;
	m_currentMountFSFile = fsDir;
	m_currentAccessType = accessType;

	std::string cmd;

	cmd.append("/usr/bin/mount -t ");
	cmd.append(m_currentMountFS);
	cmd.append(" -o ");
	cmd.append(m_currentAccessType);
	cmd.append(" ");
	cmd.append(m_currentMountFSFile);
	cmd.append(" ");
	cmd.append(m_currentMountPoint);

	if (system(cmd.c_str())) {
		return EXIT_FAILURE;
	}
	else {
		return EXIT_SUCCESS;
	}
}

int Mount::fsUmount() {
	// Checks to see if we are running as root.
	// ARoot is required to use the `umount` command
	if (geteuid() != 0) {
		return Mount::NOROOT;
	}

	// Checks to see if the mount point exists.
	if (!std::filesystem::exists(m_currentMountPoint)) {
		return Mount::MOUNTPOINTERR;
	}

	std::string cmd;

	cmd.append("/usr/bin/umount ");
	cmd.append(m_currentMountPoint);

	if (system(cmd.c_str()) == EXIT_SUCCESS) {
		return EXIT_SUCCESS;
	}
	else {
		return EXIT_FAILURE;
	}
}

std::string Mount::m_getMountString(int mountType) {
	switch (mountType) {
		case Mount::adfs:
			return "adfs";
			break;
		case Mount::affs:
			return "affs";
			break;
		case Mount::autofs:
			return "autofs";
			break;
		case Mount::cifs:
			return "cifs";
			break;
		case Mount::coda:
			return "coda";
			break;
		case Mount::coherent:
			return "coherent";
			break;
		case Mount::cramfs:
			return "cramfs";
			break;
		case Mount::debugfs:
			return "debugfs";
			break;
		case Mount::devpts:
			return "devpts";
			break;
		case Mount::efs:
			return "efs";
			break;
		case Mount::ext:
			return "ext";
			break;
		case Mount::ext2:
			return "ext2";
			break;
		case Mount::ext3:
			return "ext3";
			break;
		case Mount::ext4:
			return "ext4";
			break;
		case Mount::hfs:
			return "hfs";
			break;
		case Mount::hfsplus:
			return "hfsplus";
			break;
		case Mount::hpfs:
			return "hpfs";
			break;
		case Mount::iso9660:
			return "iso9660";
			break;
		case Mount::jfs:
			return "jfs";
			break;
		case Mount::minix:
			return "minix";
			break;
		case Mount::msdos:
			return "msdos";
			break;
		case Mount::ncpfs:
			return "ncpfs";
			break;
		case Mount::nfs:
			return "nfs";
			break;
		case Mount::nfs4:
			return "nfs4";
			break;
		case Mount::ntfs:
			return "ntfs";
			break;
		case Mount::proc:
			return "proc";
			break;
		case Mount::qnx4:
			return "qnx4";
			break;
		case Mount::ramfs:
			return "ramfs";
			break;
		case Mount::reiserfs:
			return "reiserfs";
			break;
		case Mount::romfs:
			return "romfs";
			break;
		case Mount::smbfs:
			return "smbfs";
			break;
		case Mount::squashfs:
			return "squashfs";
			break;
		case Mount::sysv:
			return "sysv";
			break;
		case Mount::tmpfs:
			return "tmpfs";
			break;
		case Mount::ubifs:
			return "ubifs";
			break;
		case Mount::udf:
			return "udf";
			break;
		case Mount::umsdos:
			return "umsdos";
			break;
		case Mount::usbfs:
			return "usbfs";
			break;
		case Mount::vfat:
			return "vfat";
			break;
		case Mount::xenix:
			return "xenix";
			break;
		case Mount::xfs:
			return "xfs";
			break;
		case Mount::xiafs:
			return "xiafs";
			break;
		default:
			return "NOTSUPPORTED";
			break;
	}
	return "NOTSUPPORTED";
}
